map2loop/interpolators.py

```diff
@@ -0,0 +1,443 @@
+from abc import ABC, abstractmethod
+from typing import Any, Union
+import beartype
+import numpy
+from numpy import ndarray
+from scipy.interpolate import Rbf, LinearNDInterpolator
+from .utils import strike_dip_vector, setup_grid
+import pandas
+
+
+class Interpolator(ABC):
+    """
+    Base Class of Interpolator used to force structure of Interpolator
+
+    Args:
+        ABC (ABC): Derived from Abstract Base Class
+    """
+
+    def __init__(self):
+        """
+        Initialiser of for Interpolator
+        """
+        self.interpolator_label = "InterpolatorBaseClass"
+
+    def type(self):
+        """
+        Getter for subclass type label
+
+        Returns:
+            str: Name of subclass
+        """
+        return self.interpolator_label
+
+    @beartype.beartype
+    @abstractmethod
+    def setup_interpolation(self, structure_data: pandas.DataFrame):
+        """
+        abstract method to setup interpolation (abstract method)
+
+        Args:
+            structure_data (pandas.DataFrame): sampled structural data
+        """
+        pass
+
+    @beartype.beartype
+    @abstractmethod
+    def setup_grid(self, bounding_box: dict):
+        """
+        abstract method to setup an XY grid (abstract method)
+
+        Args:
+            bounding_box (dict): a dictionary containing the bounding box of the map data
+        """
+        pass
+
+    @beartype.beartype
+    @abstractmethod
+    def interpolator(self, ni: Union[list, numpy.ndarray]):
+        """
+        Interpolator method
+
+        Args:
+
+            ni (int): number of points
+
+
+        Returns:
+            float: interpolated value
+        """
+
+        pass
+
+    @beartype.beartype
+    @abstractmethod
+    def interpolate(
+        self,
+        bounding_box: dict,
+        structure_data: pandas.DataFrame,
+        interpolator: Any = None,
+    ) -> Any:
+        """
+        Execute interpolate method (abstract method)
+
+        Args:
+            bounding_box (dict): a dictionary containing the bounding box of the map data
+            structure_data (pandas.DataFrame): sampled structural data
```

Comment:
Maybe specify column headers for structural data?

Reply:
I don't think it's worth it since we are going to access sampled data directly from the samples storage class, and the argument structure_data will be removed